### ISSUE TITLE
[Improvment] - Add support for Solana message signing

### ISSUE THREAD

**Issue by ffrediani** [trezor/trezor-firmware#4371]
I have been in search for a wallet that works with Trezor to manage asset in Solana blockchain. I tested most of the main ones that are known to have support for Solana (OKX, Phantom, Solfrale, Nu.FI and Backpack).
Some don't have Trezor support and some have, but partially. When trying to to sign a message a got a error "WalletSignMessageError: Not implemented"

When reached to the wallet's support they provided a comprehensive explanation:

"_This is something that Trezor needs to develop. In a nutshell, Trezor doesn't support message signing for Dapp transactions (neither does Ledger, actually, but Dapps find a workaround that they haven't yet implemented for Trezor)._"

Is it something in the radar that can be reached anytime soon ?

As a result in order to deal with Dapps it is necessary to move asset from Trezor to a Hot Wallet that can interact with Dapps on Solana which reduces the security level

**Comment by ffrediani** [comment]
Trezor Support confirmed Trezor doesn't support this feature for Solana and asked to use a Hot Wallet without Trezor.

**Comment by 0xbuts** [comment]
I have the same problem, and received a similar response from support.

I believe this is much more than an improvement, because is blocking users from using DeFi and lots of dApps on Solana securely with their Trezor. 

It would be great if a developer could add some labels to increase the perceived priority of this issue.

**Comment by monkeym4ster** [comment]
I need this feature as well. Please let me know of any updates.

**Comment by ffrediani** [comment]
@0xbuts agree. It became more than a improvement given these restrictions when interacting with certain dApps on Solana.

**Comment by matejcik** [comment]
hello all,

Solana **does not have a message signing specification**.

In the absence of a standard, developers have started to implement a so-called "message signing" by just straight up applying Solana's signature algorithm to random data. This is [kind of endorsed](https://solana.com/developers/cookbook/wallets/sign-message) by the Solana core devs.

Unfortunately, doing things this way is **extremely insecure**, because it allows **every signed message** to be understood as a **transaction** on the Solana network.
Specifically:
* you could pass a transaction to the "Sign message" call and the user would not be aware that they are signing a transaction
* a clever enough hacker may be able to devise a _readable text_ that is at the same time a valid transaction stealing your tokens or something.

Given the above, we **can not securely implement** Solana message signing on a hardware wallet until a proper message signing standard is specified.
(if a competing hardware wallet claims support for Solana message signing, I advise _extreme caution_ with what you are signing using that feature.)

**Comment by ffrediani** [comment]
What a pity, so long for this close.
@matejcik it is understandable what you said and thanks for the details, but it doesn't matter much that much if this means Trezor customers will continue to be unable to "sign messages" on certain dApps and limit their Trezor usage. That's all most will see and talk about they "Cannot use my Trezor fully".
After you explanation, I wonder what makes some dApps implement this requirement ?

If a complete secure Solana message signing can't be implemented right now, some workaround may be considered, with relevant warnings, such as your main competitors made so far, and work alongside with those who establish Solana standards to have a permanent standard later on.

**Comment by Hannsek** [comment]
We do not want to allow blind signing at all. 

Once a sane standard will be established, we'll definitely reevaluate.

**Comment by matejcik** [comment]
For completeness: we've been made aware of [this specification](https://github.com/solana-labs/solana/blob/master/docs/src/proposals/off-chain-message-signing.md) that would fulfil the requirements for Trezor implementation.
So that option is on the table, depending on how popular this spec is.

Either way, it is not on the roadmap right now.

**Comment by 0xbuts** [comment]
I'm struggling to understand the rationale behind this decision.

1) On EVM chains, some transactions on Trezor require us to blind-sign raw hexadecimal data, which is unreadable. In these cases, we have to rely entirely on third-party software to interpret what we're actually signing.
2) Trezor already offers insecure or experimental features as opt-in options.

It's perfectly understandable not to invest time in something that will likely change, given the lack of a clear standard. But given those two points, using security as the justification feels misleading.

**Comment by matejcik** [comment]
Trezor already has features that are insecure, yes, such as Ethereum data blind-signing.
These are typically grandfathered-in, and there are plans to rectify the security problems and/or deprecate said features.

That does not mean that we want to introduce **new** insecure features, _especially_ those that are fundamentally and unfixably insecure. Any engineering effort spent on introducing an insecure feature is better spent on improving security of existing features.
Under the same policy, we will not be adding arbitrary data signing, which one _could_ market as "universal support for every cryptocurrency ever"

**Comment by ffrediani** [comment]
Understandable, however you may be looking only engineering side and maybe not balancing others like product and what market expects.
If you took the decision to implement data blind-signing for Ethereum is probably because the compromise was important for product acceptance.

It seems this thing for Solana is a bit far way from be standardized, but in practice there is a growing number of Solana dApps requiring message signing and Trezor keep lacking it (even if not the ideal solution) may be as bad as introducing it as is now and improve it later.

I made the mistake to use my Trezor + Backpack for some Solana stuff and now I am locked into a situation that I can rollback and will end up having to purchase Trezor's competitor product which does it (even insecurely) just to unlock this stuff.
Also I am part of a Solana Multisig that seems to require this feature and will likely have to use Trezor's competitor to resolve this situation, as I have heard from other members how difficult it has been to deal with these Solana cases.

**Comment by sandys** [comment]
> For completeness: we've been made aware of [this specification](https://github.com/solana-labs/solana/blob/master/docs/src/proposals/off-chain-message-signing.md) that would fulfil the requirements for Trezor implementation. So that option is on the table, depending on how popular this spec is.
> 
Any chance that this will be implemented? This is now official and in the solana docs 
- https://docs.solanalabs.com/cli/examples/sign-offchain-message#protocol-specification
- https://docs.solanalabs.com/proposals/off-chain-message-signing
- https://github.com/solana-labs/solana/pull/29807

And everyone uses this (including pump.fun, etc)

**Comment by BytesOfMan** [comment]
This is necessary e.g. for completing a KYB application at Kraken. Kinda unbelievable this isn't supported.

**Comment by ffrediani** [comment]
It has been a difficult thing to convince Trezor team to do some things, some pretty obvious and no brainer sometimes.
Sometimes there is a feeling that if someone brings up something that hasn't been implemented yet, someone will respond that if it was like that someone from the team would have thought or done already.

Hopefully there is a attitude change and this necessary feature gets implemented somehow looking towards user demand and simplicity.

**Comment by Hannsek** [comment]
We are now evaluating the new approach. Please give us some time. 🙏🏻

**Comment by zktex** [comment]
Planning to implement Solana off-chain message signing in the June release. Please follow #6759 for updates.
